### PR TITLE
build: add asan checks

### PR DESCRIPTION
**Pull request by gengjiawen**
Fix https://github.com/libuv/libuv/issues/2999

### CODE REVIEW

File: .github/workflows/santilizer.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      - uses: actions/checkout@v2
+      - name: Envinfo
+        run: npx envinfo
+      - name: ASAN
+        run: |
+          mkdir build
+          cd build && cmake .. -DBUILD_TESTING=ON -DASAN=ON -DCMAKE_BUILD_TYPE=Debug
+          cmake --build .
+          ./uv_run_tests_a
+
```
Comment (bnoordhuis) [inline .github/workflows/santilizer.yml:18]: Typo in file name: `s/santilizer.yml/sanitizer.yml` (but I'd probably call it `asan.yml`.)

File: CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +30,12 @@ if(QEMU)
   add_definitions(-D__QEMU__=1)
 endif()
 
+option(ASAN "Enable code asan" OFF)
```
Comment (bnoordhuis) [inline CMakeLists.txt:33]:
```suggestion
option(ASAN "Enable AddressSanitizer (ASan)" OFF)
```
Probably a little clearer.

File: CMakeLists.txt [hunk 3]
```diff
@@ -30,6 +30,12 @@ if(QEMU)
...
+if(ASAN AND CMAKE_C_COMPILER_ID MATCHES "AppleClang|GNU|Clang")
+    set (CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=address")
+    set (CMAKE_LINKER_FLAGS_DEBUG "${CMAKE_LINKER_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=address")
```
Comment (bnoordhuis) [inline CMakeLists.txt:36]:
```suggestion
  set (CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=address")
  set (CMAKE_LINKER_FLAGS_DEBUG "${CMAKE_LINKER_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=address")
```

**Review by bnoordhuis — APPROVED** [review]
Thanks! Mostly LGTM but I suspect it's going to be hard to get the test suite to pass.

Some failures are easy to fix (e.g. `platform_output` missing a `uv_os_free_passwd()` call) but others are probably because helper processes terminate abruptly, without cleaning up first.

**Comment by gengjiawen** [comment]
Any way fix on this one, `use free(conn)` explicitly ?

```console
# got 6 bytes
# 
# =================================================================
# ==6097==ERROR: LeakSanitizer: detected memory leaks
# 
# Direct leak of 248 byte(s) in 1 object(s) allocated from:
#     #0 0x7f0118355b40 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb40)
#     #1 0x561ef53af2c4 in ipc_on_connection_tcp_conn /home/runner/work/libuv/libuv/test/test-ipc.c:704
#     #2 0x561ef547f9ca in uv__server_io /home/runner/work/libuv/libuv/src/unix/stream.c:570
#     #3 0x561ef54973d2 in uv__io_poll /home/runner/work/libuv/libuv/src/unix/linux-core.c:461
#     #4 0x561ef5454842 in uv_run /home/runner/work/libuv/libuv/src/unix/core.c:385
#     #5 0x561ef53b0cd2 in ipc_helper_tcp_connection /home/runner/work/libuv/libuv/test/test-ipc.c:822
#     #6 0x561ef533ad1e in maybe_run_test /home/runner/work/libuv/libuv/test/run-tests.c:105
#     #7 0x561ef533a95d in main /home/runner/work/libuv/libuv/test/run-tests.c:67
#     #8 0x7f0117881b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
# 
```

File: test/test-platform-output.c [hunk 4]
```diff
@@ -171,5 +171,6 @@ TEST_IMPL(platform_output) {
   printf("  version: %s\n", uname.version);
   printf("  machine: %s\n", uname.machine);
 
+  uv_os_free_passwd(&pwd);
```
Comment (cjihrig) [inline test/test-platform-output.c:174]: For consistency with the other free operations in this test, can you move this up to line 158.

**Comment by bnoordhuis** [comment]
> Any way fix on this one, use `free(conn)` explicitly ?

Not until `close_cb` is invoked... but you probably can't just stick a `free(handle)` in there because it's used by more `uv_close()` calls. Probably easiest to add a new callback:
<details>

```diff
diff --git a/test/test-ipc.c b/test/test-ipc.c
index 39ef4f11..eaa47c41 100644
--- a/test/test-ipc.c
+++ b/test/test-ipc.c
@@ -587,6 +587,12 @@ static void close_cb(uv_handle_t* handle) {
 }
 
 
+static void close_and_free_cb(uv_handle_t* handle) {
+  close_cb_called++;
+  free(handle);
+}
+
+
 static void conn_notify_write_cb(uv_write_t* req, int status) {
   uv_close((uv_handle_t*)&tcp_server, close_cb);
   uv_close((uv_handle_t*)&channel, close_cb);
@@ -721,7 +727,7 @@ static void ipc_on_connection_tcp_conn(uv_stream_t* server, int status) {
                     on_tcp_child_process_read);
   ASSERT_EQ(r, 0);
 
-  uv_close((uv_handle_t*)conn, close_cb);
+  uv_close((uv_handle_t*)conn, close_and_free_cb);
 }
 
 
```

</details>

**Comment by gengjiawen** [comment]
> > Any way fix on this one, use `free(conn)` explicitly ?
> 
> Not until `close_cb` is invoked... but you probably can't just stick a `free(handle)` in there because it's used by more `uv_close()` calls. Probably easiest to add a new callback:

Fixed. There is still 6 test cases failed related to file
```console
not ok 63 - fs_event_watch_file_twice
# exit code 134
not ok 174 - pipe_connect_to_file
# exit code 134
...
```

**Comment by bnoordhuis** [comment]
> Maybe performance downgrade caused by asan ?

Yes, I expect so, at least for the `fs_event_*` tests. They're pretty timing sensitive.

**Comment by gengjiawen** [comment]
Now failed with
```
ok 32 - error_message # SKIP SUMMARY: AddressSanitizer: 23 byte(s) leaked in 1 allocation(s).
ok 347 - tty_pty # SKIP SUMMARY: AddressSanitizer: 192 byte(s) leaked in 2 allocation(s).
##[error]Process completed with exit code 1.
```

**Comment by gengjiawen** [comment]
> Now [failed with](https://github.com/libuv/libuv/pull/2998/checks?check_run_id=1144305977)
> 
> ```
> ok 32 - error_message # SKIP SUMMARY: AddressSanitizer: 23 byte(s) leaked in 1 allocation(s).
> ok 33 - fork_fs_events_child
> ok 347 - tty_pty # SKIP SUMMARY: AddressSanitizer: 192 byte(s) leaked in 2 allocation(s).
> ok 348 - udp_alloc_cb_fail
> ##[error]Process completed with exit code 1.
> ```
> 
> A little confused with ok status and skip summary (since asan find leaks).

I skipped the two test. CI should be green.

**Review by cjihrig — APPROVED** [review]

**Comment by gengjiawen** [comment]
@bnoordhuis @cjihrig Can you merge this PR ? Thanks.

**Review by bnoordhuis — APPROVED** [review]
Thanks Jiawen, LGTM. I'll merge this in a moment.

**Comment by bnoordhuis** [comment]
Landed in 97a90330.

**Pull request closed without merging**